### PR TITLE
[9.x] Fix race condition in locks issued by the file cache driver

### LINKED ISSUE #43627: Atomic locks do not provide reliable mutual exclusion when using the file cache driver
- Laravel Version: 9.23.0
- PHP Version: 8.0.15
- Database Driver & Version: none

### Description:
The locks returned by [FileStore](https://github.com/laravel/framework/blob/053840f579cf01d353d81333802afced79b1c0af/src/Illuminate/Cache/FileStore.php) are instances of the class [CacheLock](https://github.com/laravel/framework/blob/053840f579cf01d353d81333802afced79b1c0af/src/Illuminate/Cache/CacheLock.php). CacheLock's implementation of the acquire method is not atomic when the lock timeout is 0 (the default):

https://github.com/laravel/framework/blob/053840f579cf01d353d81333802afced79b1c0af/src/Illuminate/Cache/CacheLock.php#L30-L50

When the lock timeout is 0, first a `get` and then a `put` is performed. When combined, these operations are not atomic and thus multiple actors can enter the critical section protected by the lock. The possibility for race conditions was already discussed in the PR https://github.com/laravel/framework/pull/35139 that introduced the change.

#### Proposed quick fix
A quick fix would be to always call `$this->store->add` even if  `$this->seconds <= 0` because the store can handle 0 timeouts just fine:

```php
public function acquire() 
{ 
    return $this->store->add($this->name, $this->owner, $this->seconds); 
}
```

There is no need to even check that the `add` method exists because it is the stores responsibility to return a compatible lock instance.

#### Proposed proper fix
The proper fix would be to write a custom lock implementation for each store. There already exists [DatabaseLock](https://github.com/laravel/framework/blob/053840f579cf01d353d81333802afced79b1c0af/src/Illuminate/Cache/DatabaseLock.php), [RedisLock](https://github.com/laravel/framework/blob/053840f579cf01d353d81333802afced79b1c0af/src/Illuminate/Cache/RedisLock.php), [DynamoDbLock](https://github.com/laravel/framework/blob/053840f579cf01d353d81333802afced79b1c0af/src/Illuminate/Cache/DynamoDbLock.php) etc. which can guarantee mutual exclusion because their implementation is tailored to the specific store. There is no good way to write an abstract lock implementation that [CacheLock](https://github.com/laravel/framework/blob/053840f579cf01d353d81333802afced79b1c0af/src/Illuminate/Cache/CacheLock.php) attempts to be because the [Store](https://github.com/laravel/framework/blob/053840f579cf01d353d81333802afced79b1c0af/src/Illuminate/Contracts/Cache/Store.php) contract does not provide the necessary atomic primitives. Calling an undocumented `add` method on the store, which may or may not be atomic, seems like the wrong thing to do.

### Steps To Reproduce:

I created a a repo that can pretty reliably reproduce the race condition:
https://github.com/hbgl/laravel-test-locks

Instructions:
```bash
git clone https://github.com/hbgl/laravel-test-locks
cd laravel-test-locks
composer install
php artisan app:test-locks --workers=2 --iterations=1000 --cache-driver=file
```

### PR DESCRIPTION
**Pull request by bytestream**
Fixes https://github.com/laravel/framework/issues/43627

Before:

```bash
root@fa0e2058d0ad:/app# php artisan app:test-locks --workers=2 --iterations=1000 --cache-driver=file
Running 1000 iterations with 2 workers using the cache driver 'file' for locking.
Process 1/2 started.
Process 2/2 started.
Process 1/2 finished.
Process 2/2 finished.
Process 1/2 started at 0 and ended at 1938.
Process 2/2 started at 0 and ended at 1932.
Counted only to 1938 instead of the expected 2000.
```

After:

```bash
root@fa0e2058d0ad:/app# php artisan app:test-locks --workers=2 --iterations=1000 --cache-driver=file
Running 1000 iterations with 2 workers using the cache driver 'file' for locking.
Process 1/2 started.
Process 2/2 started.
Process 1/2 finished.
Process 2/2 finished.
Process 1/2 started at 0 and ended at 1922.
Process 2/2 started at 26 and ended at 2000.
Successfully counted to 2000.

```

### CODE REVIEW

**Comment by taylorotwell** [comment]
Please explain how this changes anything? The FileStore already has an `add` method so it would already be falling into this conditional block anyways?

**Comment by bytestream** [comment]
@taylorotwell the problem is when `$this->seconds = 0` it falls to the `get` / `put` combination which isn't atomic. The `add` method in the file driver is able to handle when `$seconds = 0`, and treats it as cache forever

**Comment by bytestream** [comment]
@taylorotwell [362fc83](https://github.com/laravel/framework/pull/46011/commits/362fc838c060aec78d992b3538fb4ae5c20e95f9) is a cleaner way of handling it while maintaining BC

**Comment by taylorotwell** [comment]
Are other drivers not able to handle `$seconds = 0`?

**Comment by taylorotwell** [comment]
I also don't agree your second commit is necessarily better. The first was simpler?

**Comment by bytestream** [comment]
@taylorotwell `CacheLock` and `HasCacheLock` are only actually used by the file driver. They're a legacy from https://github.com/laravel/framework/pull/35139 which originally added APC driver support as well, but was removed before it was merged. The simplest fix may just be the one that was originally suggested in https://github.com/laravel/framework/pull/43661. If it were me though, I'd remove the `HasCacheLock` and rename `CacheLock` to `FileLock`. That's clearer and also follows the way other drivers are implemented.

**Comment by taylorotwell** [comment]
Sorry - I'm not following at all but I don' have time to dig deeper atm. 

1) Can other drivers not handle $seconds = 0? 

2) Why is this not the simplest fix and easier?

CleanShot 2023-02-09 at 16 55 37@2x

**Comment by taylorotwell** [comment]
For what it's worth: I've updated the docs to not even show an example of not providing an expiration time (in seconds) to the lock methods.

**Comment by bytestream** [comment]
> Can other drivers not handle $seconds = 0?

TLDR; not relevant as `CacheLock` is **only** used by the `FileStore`. To directly answer the question though:

* APC - :x: , missing `add` method
* Array - :x: , missing `add` method
* Database - :ok: , providing `$seconds = 0` to `add()` results in immediate expiry, but `DatabaseLock` uses expiry of 1 day (86400 seconds) if `$seconds = 0`
* DynamoDB - :ok: , providing `$seconds = 0` to `add()` results in immediate expiry, but `DynamoDbLock` uses expiry of 1 day (86400 seconds) if `$seconds = 0`
* File - :ok: , expiry is set to `9999999999`
* Memcached - :ok: , `0` means never expire - https://www.php.net/manual/en/memcache.add.php
* Null - :x: , missing `add` method
* Redis - :ok: , providing `$seconds = 0` to `add()` expires after 1 second, but `RedisLock` correctly handles `$seconds = 0`

> Why is this not the simplest fix and easier?

Simplest fix in the sense that it's a 1 line change, but given `CacheLock` is **only** used by the `FileStore` it really makes sense to refactor the usage. https://github.com/laravel/framework/pull/43661 is the simplest fix IMO without any additional refactoring, and **has no breaking changes / consequences.**
